File: app/src/main/java/kr/pe/lightvil/repositories/CallLogRepository.java
package kr.pe.lightvil.repositories;
import android.content.Context;

import java.util.List;
import java.util.Map;

import kr.pe.lightvil.daos.CallLogDAO;
import kr.pe.lightvil.models.CallLog;
import kr.pe.lightvil.models.DialogMessage;
import kr.pe.lightvil.models.EyePhoneDatabase;

public class CallLogRepository {
    private CallLogDAO callLogDAO;


    public CallLogRepository(Context context) {
        this.callLogDAO = EyePhoneDatabase.getDatabase(context).getCallLogDAO();
    }

    public int getLogCount() {
        return this.callLogDAO.getLogCount();
    }

    public List<CallLog> findAllCallLogs() {
        return callLogDAO.findAll();
    }

    public CallLog findCallLogById(Long callLogId) {
        return callLogDAO.findById(callLogId);
    }

    public Map<CallLog, List<DialogMessage>> findCallLogWithMessagesById(Long callLogId) {
        return callLogDAO.findWithMessagesById(callLogId);
    }

    public List<DialogMessage> findMessagesById(Long callLogId) {
       return callLogDAO.findMessagesByCallLogId(callLogId);
    }

    public void insert(CallLog callLog, List<DialogMessage> messages) {
        EyePhoneDatabase.executeWrite(() -> {
            long newId = callLogDAO.insertCallLogWithMessages(callLog, messages);
            callLog.setId(newId);
        });
    }

    public void insert(CallLog callLog) {
        EyePhoneDatabase.executeWrite(() -> {
            long newId = callLogDAO.insertCallLog(callLog);
            callLog.setId(newId);
        });
    }
}

File: app/src/main/kotlin/kr/pe/lightvil/repositories/CallLogRepository.kt
package kr.pe.lightvil.repositories

import android.content.Context
import androidx.paging.PagingSource
import kr.pe.lightvil.daos.CallLogDao
import kr.pe.lightvil.models.*

class CallLogRepository(val context: Context)  {
    private val  dao : CallLogDao by lazy { EyePhoneDatabase.getInstance(context)!!.callLogDao() }

    fun getCallLogs(query : String) : CallLogPagingSource {
        return dao.getCallLogs(query)
    }

    fun getMessageByCallLogId(callLogId: Long, query: String) : CallLogMessagePagingSource {
        return dao.getMessagesByCallLogId(callLogId, query)
    }

    suspend fun insert(callLog: CallLog) : Long {
        val newId: Long = dao.insertCallLog(callLog)
        callLog.id = newId
        return newId
    }

    suspend fun insert(callLog: CallLog, messages : List<CallLogMessage>) : Long {
        return dao.insertCallLogWithMessages(callLog, messages)!!
    }

    suspend fun delete(callLog: CallLog) {
        dao.deleteCallLogAndMessages(callLog)
    }
}